### ISSUE TITLE
libsubprocess: Do not assume EOF on stream when child has exited

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#2228]
In #2211, @garlick found circumstances where stdout was not consistent between tasks even though the task output should be identical.

After a lot of debugging, it was determined that some subprocesses appear to send an EOF to a callback / identify EOF on a stream if 1) the buffer is empty and 2) the subprocess has exited, BUT an EOF had not actually been identified via the flux buffer watcher (i.e. with `flux_buffer_read_watcher_create()`).

I believe now this logic is flawed.  Under a more traditional fork, wait, read from file description pattern, this logic is probably safe/correct (i.e. spin on `read()` until you hit zero bytes).  However, in the `libsubprocess` library, this is probably unsafe given everything is done via the reactor (i.e. spin on `flux_buffer_bytes()` until zero bytes is not safe).   When the subprocess exits, leftover stdout & EOF is flushed, copied / detected in the buffer, a callback indicating data / EOF is in the buffer, and the child process has exited are all handled by the reactor.  And the order of the reactor events may not correspond to what one might normally think.

I'm eyeing this code in `local.c`.

```
        output_cb (c->p, c->name);

        if (c->p->state == FLUX_SUBPROCESS_EXITED && !c->eof_sent_to_caller) {

            if (!(fb = flux_buffer_read_watcher_get_buffer (w))) {
                flux_log_error (c->p->h, "flux_buffer_read_watcher_get_buffer");
                return;
            }

            if (!flux_buffer_bytes (fb)) {

                output_cb (c->p, c->name);

                c->eof_sent_to_caller = true;
                eof_set = true;
                c->p->channels_eof_sent++;
            }
        }
```

Here, a callback to the user (via `output_cb`, which will call the `on_stdout` callback set by the user) is called.  In it the user could read all of the data in the buffer.  Afterwards, it is assumed that if the subprocess has exited and the buffer is empty, that we must have read EOF on the stream.  But that may not necessarily be true.

It might be ok to simply remove this block of code and always wait for the "real" EOF to come from the read buffer reactor.  But before doing so got to 1) try and make some reproducers, 2) see if this exists on the remote subprocess side of code (and see if I can make a reproducer on that side if necessary).

**Comment by chu11** [comment]
Reviewing code, generally speaking this doesn't exist on the remote execution part of the libsubprocess, except that b/c the detection of EOF is flawed locally, it could "get into" the remote execution part of this equation.

**Comment by chu11** [comment]
after a lot of attempts to create a simple reproducer, it seems that something akin to what @grondo did in #2211 is the only thing that seems to reproduce.  The keys being 1) the output should be > 4096 bytes, 2) there should be a lot of lines of output, 3) you need to run with `taskset -c 0`.  AFAICT, launching many local processes in parallel just doesn't seem to do the trick.

On another note, it does appear that just removing the code listed above fixes the task output issue discovered in #2211 (atleast using the `taskset -c 0` reproducer).

**Comment by chu11** [comment]
well whatever was reproducing for me isn't reproducing anymore, I guess it takes some special stars to align :-(
